scapy/contrib/automotive/someip.py

```diff
@@ -129,23 +142,76 @@ class SOMEIP(Packet):
             RET_E_MALFORMED_MSG: "E_MALFORMED_MESSAGE",
             RET_E_WRONG_MESSAGE_TYPE: "E_WRONG_MESSAGE_TYPE",
         }),
+        ConditionalField(BitField("offset", 0, 28),
+                         lambda pkt: SOMEIP._is_tp(pkt)),
+        ConditionalField(BitField("res", 0, 3),
+                         lambda pkt: SOMEIP._is_tp(pkt)),
+        ConditionalField(BitField("more_seg", 0, 1),
+                         lambda pkt: SOMEIP._is_tp(pkt))
     ]
 
     def post_build(self, pkt, pay):
         length = self.len
-        if (length is None):
-            length = self.LEN_OFFSET + len(pay)
-            pkt = pkt[:4] + struct.pack("!I", length) + pkt[8:]
+        if length is None:
+            if SOMEIP._is_tp(self):
+                length = SOMEIP.LEN_OFFSET_TP + len(pay)
+                pkt = pkt[:4] + struct.pack("!I", length) + pkt[8:]
+            else:
+                length = SOMEIP.LEN_OFFSET + len(pay)
+                pkt = pkt[:4] + struct.pack("!I", length) + pkt[8:]
         return pkt + pay
 
     def answers(self, other):
         if other.__class__ == self.__class__:
             if self.msg_type in [SOMEIP.TYPE_REQUEST_NO_RET,
-                                 SOMEIP.TYPE_REQUEST_NORET_ACK]:
+                                 SOMEIP.TYPE_REQUEST_NORET_ACK,
+                                 SOMEIP.TYPE_NOTIFICATION,
+                                 SOMEIP.TYPE_TP_REQUEST_NO_RET,
+                                 SOMEIP.TYPE_TP_NOTIFICATION]:
                 return 0
             return self.payload.answers(other.payload)
         return 0
 
+    @staticmethod
+    def _is_tp(pkt):
+        """Returns true if pkt is using SOMEIP-TP, else returns false."""
+
+        tp = [SOMEIP.TYPE_TP_REQUEST, SOMEIP.TYPE_TP_REQUEST_NO_RET,
+              SOMEIP.TYPE_TP_NOTIFICATION, SOMEIP.TYPE_TP_RESPONSE,
+              SOMEIP.TYPE_TP_ERROR]
+        if isinstance(pkt, Packet):
+            return pkt.msg_type in tp
+        else:
+            return pkt[15] in tp
+
+    def fragment(self, fragsize=1392):
```

Comment:
You can ignore the issue on this one. Parent has `def fragment(self, *args, **kargs):` but in this case it’s fine if you don’t match it

Reply:
Ok, thanks.